scylla-cdc/src/reader.rs

```diff
@@ -0,0 +1,340 @@
+use std::cmp::{max, min};
+use std::error::Error;
+use std::sync::Arc;
+use std::time;
+use std::time::SystemTime;
+
+use chrono;
+use futures::Future;
+use scylla::frame::value::Timestamp;
+use scylla::transport::iterator::RowIterator;
+use scylla::Session;
+use tokio::time::sleep;
+
+use crate::stream_id::StreamID;
+
+pub struct StreamReader {
+    session: Arc<Session>,
+    stream_id_vec: Vec<StreamID>,
+    lower_timestamp: chrono::Duration,
+    window_size: chrono::Duration,
+    safety_interval: chrono::Duration,
+    upper_timestamp: tokio::sync::Mutex<Option<chrono::Duration>>,
+    sleep_interval: time::Duration,
+}
+
+impl StreamReader {
+    pub fn new(
+        session: &Arc<Session>,
+        stream_ids: Vec<StreamID>,
+        start_timestamp: chrono::Duration,
+        window_size: chrono::Duration,
+        safety_interval: chrono::Duration,
+        sleep_interval: time::Duration,
+    ) -> StreamReader {
+        StreamReader {
+            session: session.clone(),
+            stream_id_vec: stream_ids,
+            lower_timestamp: start_timestamp,
+            window_size,
+            safety_interval,
+            upper_timestamp: Default::default(),
+            sleep_interval,
+        }
+    }
+
+    pub async fn set_upper_timestamp(&mut self, new_upper_timestamp: chrono::Duration) {
+        let mut guard = self.upper_timestamp.lock().await;
+        *guard = Some(new_upper_timestamp);
+    }
+
+    pub async fn fetch_cdc<Fut>(
+        &mut self,
+        keyspace: String,
+        table_name: String,
+        after_fetch_callback: fn(RowIterator) -> Fut,
```

Comment:
I don't think a raw function pointer will be enough for tests. `fn(RowIterator) -> Fut` is a raw function type - which means that it doesn't allow capturing state, just like function pointers in C. In tests, you will probably want to record the data which was provided to the callback and perform some assertions on that data (e.g. that the stream reader managed to read a particular amount of rows). In order to do that, the test function and the closure provided by the test will have to share some variables - for example, the callback will append to a Vec, and after `fetch_cdc` finishes, the test function will analyze the data.

Additionally, I'd suggest changing the `fetch_cdc` so that it iterates over the stream itself and passes rows to the callback.